build.sbt

```diff
@@ -35,67 +38,46 @@ def flattenTasks[A](tasks: Seq[Def.Initialize[Task[A]]]): Def.Initialize[Task[Se
-  val benchBase = loader.loadClass("org.renaissance.RenaissanceBenchmark")
+def listBenchmarks(project: String, classpath: Seq[File]) = {
+  //
+  // Load the benchmark base class and create a class loader for the project
+  // with the class loader of the base class as its parent. This will allow
+  // us to use core classes here.
+  //
+  val benchBase = classOf[RenaissanceBenchmark]
+  val urls = classpath.map(_.toURI.toURL).toArray
+  val loader = new URLClassLoader(urls, benchBase.getClassLoader)
   val excludePattern = Pattern.compile("org[.]renaissance(|[.]harness|[.]util)")
-  val result = new mutable.ArrayBuffer[(String, String, String, String, Int)]
-  for (jar <- classpath) {
-    val jarFile = new JarFile(jar)
-    val enumeration = jarFile.entries()
-    while (enumeration.hasMoreElements) {
-      val entry = enumeration.nextElement()
+
+  //
+  // Scan all JAR files for classes in the org.renaissance package implementing
+  // the benchmark interface and collect metadata from class annotations.
+  //
+  val result = new mutable.ArrayBuffer[BenchmarkInfo]
+  for (jarFile <- classpath.map(jar => new JarFile(jar))) {
+    for (entry <- JavaConverters.enumerationAsScalaIterator(jarFile.entries())) {
       if (entry.getName.startsWith("org/renaissance") && entry.getName.endsWith(".class")) {
-        val name = entry.getName
+        val benchClassName = entry.getName
           .substring(0, entry.getName.length - ".class".length)
           .replace("/", ".")
-        val clazz = loader.loadClass(name)
+        val clazz = loader.loadClass(benchClassName)
 
         val isEligible =
           !excludePattern.matcher(clazz.getPackage.getName).matches() &&
             benchBase.isAssignableFrom(clazz)
         if (isEligible) {
-          // Can we PLEASE have a reasonable logging support in SBT?
-          // And NOT the streams.value or sLog.value that cannot be used here?
-          // It's a turing-complete build system and we can't even log conveniently!
-          println("eligible benchmark: " + clazz.getName)
-
-          val instance = clazz.getDeclaredConstructor().newInstance()
-          val distro = clazz.getMethod("distro").invoke(instance).toString
-          val licenses = clazz
-            .getMethod("licenses")
-            .invoke(instance)
-            .asInstanceOf[Array[Object]]
-            .map(x => x.toString)
-            .mkString(",")
-          val description = clazz.getMethod("description").invoke(instance).toString
-          val reps =
-            Integer.parseInt(clazz.getMethod("defaultRepetitions").invoke(instance).toString)
-          result += ((kebabCase(clazz.getSimpleName), distro, licenses, description, reps))
+          // Print info to see what benchmarks are picked up by the build.
+          val benchClass = clazz.asSubclass(benchBase)
+          val info = new BenchmarkInfo(benchClass)
+          info.print()
```

Comment:
Is this intentional or a debugging left-over? Of course, not a big issue.

Reply:
Intentional. If logging was conveniently available (it's not, or I can't find it), I would use it. When playing with the build, it was actually a welcome sight to see what benchmarks were picked up and with what metadata. Compared to the mostly useless output of SBT, this at least tells me that build system found the most important things -- the benchmarks ;-)

I'll have a look if logging could be accessed more conveniently from BenchmarkInfo class (instead of build.sbt), so that it can be just logged.

Reply:
Not really, only if the logger is somehow passed from a task to the class that might need it. The antics involved look like standing on one hand and using the feet to juggle three clay pots filled with nitroglycerine.